protocols/identify/src/lib.rs

```diff
@@ -229,15 +248,52 @@ impl<T: AddrManager> ServiceProtocol for IdentifyProtocol<T> {
                     }
                 } else {
                     trace!("received observed address: {}", addr);
-                    info.observed_addr = Some(addr.clone());
-                    // TODO how can we trust this address?
-                    if self
-                        .addr_mgr
-                        .add_observed_addr(&info.peer_id, addr.clone())
-                        .is_disconnect()
-                    {
-                        service.disconnect(session.id);
+
+                    // TODO: this address shoulde be reachable
+                    // Add transform observed address to local listen address list
+                    if let Some(socket_addr) = multiaddr_to_socketaddr(&addr) {
+                        let observed_ip = socket_addr.ip();
+
+                        // replace observed address's port part
+                        if self
+                            .listen_addrs
+                            .iter()
+                            .filter_map(|listen_addr| multiaddr_to_socketaddr(listen_addr))
+                            .map(|socket_addr| socket_addr.ip())
+                            .all(|listen_ip| listen_ip != observed_ip)
+                        {
+                            // TODO: Choose a random port from current local listen addresses
```

Comment:
Why choose a random port? it's not reasonable, should discover address from all listened port.

Reply:
How about listens:
* `192.168.0.1:8000`
* `137.34.23.4:8989`
* `167.33.13.1:7788`

Reply:
We can filter these addrs with `is_global`

Reply:
No, we can not, just the first address is not global. The machine may have several network cards.